google/cloud/internal/populate_common_options_test.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/populate_common_options.h"
+#include "google/cloud/common_options.h"
+#include "google/cloud/internal/user_agent_prefix.h"
+#include "google/cloud/testing_util/scoped_environment.h"
+#include "absl/types/optional.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+using ::google::cloud::testing_util::ScopedEnvironment;
+using ::testing::Contains;
+using ::testing::ElementsAre;
+using ::testing::IsEmpty;
+
+TEST(PopulateCommonOptions, Simple) {
+  // Unset all the relevant environment variables
+  ScopedEnvironment user("GOOGLE_CLOUD_CPP_USER_PROJECT", absl::nullopt);
+  ScopedEnvironment tracing("GOOGLE_CLOUD_CPP_ENABLE_TRACING", absl::nullopt);
+  auto actual =
+      PopulateCommonOptions(Options{}, {}, {}, "default.googleapis.com");
+  EXPECT_TRUE(actual.has<EndpointOption>());
+  EXPECT_TRUE(actual.has<UserAgentProductsOption>());
+  EXPECT_THAT(actual.get<UserAgentProductsOption>(),
+              Contains(UserAgentPrefix()));
+  EXPECT_TRUE(actual.has<TracingComponentsOption>());
+  EXPECT_THAT(actual.get<TracingComponentsOption>(), IsEmpty());
+  EXPECT_FALSE(actual.has<UserProjectOption>());
+}
+
+TEST(PopulateCommonOptions, Endpoint) {
+  auto const empty = Options{};
+  auto const with_ep = Options{}.set<EndpointOption>("with-ep");
+  struct TestCase {
+    std::string name;
+    Options initial;
+    absl::optional<std::string> default_env;
+    absl::optional<std::string> emulator_env;
+    std::string expected;
+  } test_cases[] = {
+      {"empty-0", empty, absl::nullopt, absl::nullopt, "default"},
+      {"empty-1", empty, absl::nullopt, "emulator", "emulator"},
+      {"empty-2", empty, "default", absl::nullopt, "default"},
+      {"empty-3", empty, "default", "emulator", "emulator"},
+      {"with-ep-0", with_ep, absl::nullopt, absl::nullopt, "with-ep"},
+      {"with-ep-1", with_ep, absl::nullopt, "emulator", "emulator"},
+      {"with-ep-2", with_ep, "default", absl::nullopt, "with-ep"},
+      {"with-ep-3", with_ep, "default", "emulator", "emulator"},
+  };
```

Comment:
maybe add one more case where the emulator endpoint is `""` instead of `absl::nullopt` ?

Reply:
Done.